### ISSUE TITLE
Is noinit pragma working? Or, how do I check that noinit is working?

### ISSUE THREAD

**Issue by kaushalmodi** [nim-lang/Nim#7852]
Hello,

I have been playing with little Nim code snippets, and one snippet where I have got stuck since about a week is understanding the `noInit` pragma.

I have this little snippet:

```nim
proc bar() =
  var a_noinit {.noInit.}: array[0.. 4, char]
  var a_init: array[0.. 4, char]
  echo "Is the uninitialized variable 'a_noinit' declared? ", declared(a_noinit)
  echo "Value of uninitialized variable 'a_noinit': ", a_noinit
  echo "Value of auto-initialized variable 'a_init': ", a_init
bar()
block foo:
  var a_noinit {.noInit.}: array[0.. 4, char]
  var a_init: array[0.. 4, char]
  echo "Is the uninitialized variable 'a_noinit' declared? ", declared(a_noinit)
  echo "Value of uninitialized variable 'a_noinit': ", a_noinit
  echo "Value of auto-initialized variable 'a_init': ", a_init
```

But it's outputting:

```
Is the uninitialized variable 'a_noinit' declared? true
Value of uninitialized variable 'a_noinit': ['\x00', '\x00', '\x00', '\x00', '\x00']
Value of auto-initialized variable 'a_init': ['\x00', '\x00', '\x00', '\x00', '\x00']
Is the uninitialized variable 'a_noinit' declared? true
Value of uninitialized variable 'a_noinit': ['\x00', '\x00', '\x00', '\x00', '\x00']
Value of auto-initialized variable 'a_init': ['\x00', '\x00', '\x00', '\x00', '\x00']
```

As you see, the values of uninitialized `a_noinit`, and the auto-initialized `a_init` is the exact same.. so how I understand what the `{.noInit.}` pragma is doing? 

---

I am using Nim devel built using https://github.com/nim-lang/Nim/commit/072cd073ba40d05ace580292735b5b385544d3e7.

/cc @data-man As you see, I am trying to use this pragma on local vars. But still I don't see any difference with or without using it. So may be the question is.. when you do use this pragma? What's its effect?

**Comment by data-man** [comment]
@kaushalmodi 
Try this:
```nim
proc a() =
  var
    ai {.noinit.}: array[3, int]
    ac {.noinit.}: array[3, char]

  echo ai
  echo ac

a()
```
What you see?

Maybe it's bug of echo.

**Comment by kaushalmodi** [comment]
Hmm.. I see something random with the int array..

```nim
proc a() =
  var
    ai {.noinit.}: array[3, int]
    ac {.noinit.}: array[3, char]
    ai_init: array[3, int]
    ac_init: array[3, char]

  echo ai
  echo ac
  echo ai_init
  echo ac_init

a()
```

Output:

```
[0, 140731051390112, 0]
['\x00', '\x00', '\x00']
[0, 0, 0]
['\x00', '\x00', '\x00']
```

**Comment by data-man** [comment]
Run multiple times.

**Comment by kaushalmodi** [comment]
Actually I am coming to a strange conclusion.. "`int` seems to be *unlocking* the uninit power for `char`".

Run the below multiple times.. the `ac_noinit` array will always be all zeros. Now uncomment the `ai_noinit`.. both `ai_noinit` and `ac_noinit` will start randomizing i.e. showing up uninitialized.

```
proc bar() =
  var 
    # ai_noinit {.noInit.}: array[5, int]
    ac_noinit {.noInit.}: array[5, char]
    a_init: array[5, char]
  # echo "Value of uninitialized variable 'ai_noinit': ", ai_noinit
  echo "Value of uninitialized variable 'ca_noinit': ", ac_noinit
  echo "Value of auto-initialized variable 'a_init': ", a_init
bar()
```

**Comment by data-man** [comment]
New version :)
```nim
proc myecho(a: openarray[char]) =
  var
    s: string

  for c in a:
    s.add $c.ord & ", "

  echo s[0 .. ^3]

proc bar() =
  var
    ac3_noinit {.noInit.}: array[3, char]
    ac5_noinit {.noInit.}: array[5, char]
    ac5_init: array[5, char]

  echo "Value of uninitialized variable 'ac3_noinit': ", ac3_noinit
  echo "Value of uninitialized variable 'ac5_noinit': ", ac5_noinit
  echo "Value of auto-initialized variable 'ac5_init': ", ac5_init
  myecho ac3_noinit
  myecho ac5_noinit
  myecho ac5_init

bar()
```
ac3_noinit is uninitialized for me.
Very interesting.

**Comment by kaushalmodi** [comment]
> ac3_noinit is uninitialized for me.

Hmm.. but both ac3_noinit and ac5_noinit are getting initialized for me to 0 (multiple runs).

What more, even if I change the ac3 array type to int, both ac3 and ac5 still get initialized (multiple runs). Only when I increase the length of that int array from 3 to 5, I get both to remain uninitialized.

Looks like the behavior depends on the machine, cache history, etc.?

**Comment by kaushalmodi** [comment]
That said, can you confirm that even local variables, if not arrays, always get initialized? I had a local var of `float` type without any value assigned, and it still got a value of 0. But as soon as I made it an array, it started printing randomined/uninitialized values.

**Comment by data-man** [comment]
```nim
proc bar() =
  var
    i32: int32
    i64: int64
    f32: float32
    f64: float64

    i32ni {.noInit.}: int32
    i64ni {.noInit.}: int64
    f32ni {.noInit.}: float32
    f64ni {.noInit.}: float64

  echo "i32 == ", i32
  echo "i64 == ", i64
  echo "f32 == ", f32
  echo "f64 == ", f64

  echo "i32ni == ", i32ni
  echo "i64ni == ", i64ni
  echo "f32ni == ", f32ni
  echo "f64ni == ", f64ni

bar()
```

**Comment by kaushalmodi** [comment]
Only `f64ni` seems to not get initialized. It might seem that I am trolling.. so here's the proof.. I am evaluating the whole code snippet using a single key binding in Emacs/Org mode.. you can see the `f64ni` value changing each time, but every other `*ni` var stays at zero.

https://i.imgur.com/Ertdv7Q.gifv

**Comment by kaushalmodi** [comment]
Here's another gif to show my "int unlocks uninit power deduction":

https://i.imgur.com/4mPDc3a.gifv

For first half of the GIF, the int, char, bool, float arrays, all show the expected behavior. Then comment out the int array, and then only the `char` and `bool` arrays stop randomizing.. they stay at their auto-initialization values inspite of the `noinit` pragma.

**Comment by kaushalmodi** [comment]
> Variables not in the scope of a proc or .global are always initialized by C. Local variables are not initialized if you use .noInit and this one of these rare cases where checking the produced C code helps.

Yes, all of that is known by the end of this thread, thanks to @data-man. Can that please be documented somewhere? 

The main issue is revealed in the inconsistencies that I demonstrate in the GIFs that I share in my last 2 comments in this thread. 

I have spent hours trying to understand those inconsistencies in the `noinit` behavior. I'd appreciate if some justice is done to that.

**Comment by Araq** [comment]
What's there to be documented? The compiler is free to produce code that initializes the data to any value when you use ``.noInit``, including binary zero.

**Comment by data-man** [comment]
@Araq 
From https://nim-lang.org/docs/manual.html#statements-and-expressions-var-statement

> Var statements declare new local and global variables and initialize them.
> If an initializer is given the type can be omitted: the variable is then of the same type as the initializing expression. **Variables are always initialized with a default value if there is no initializing expression**. The default value depends on the type and is **always a zero in binary**.
> **The implicit initialization can be avoided** for optimization reasons **with the noinit pragma**


This should be improved.

**Comment by data-man** [comment]
Я уже исправил неправильный пример, но на сайте неправильная версия.

**Comment by Araq** [comment]
I disagree, it's very clear, "can be avoided" and where do we write "arbitrary contents cannot be zero" (!) anyway.

**Comment by data-man** [comment]
@Araq 

> Variables are always initialized with a default value

Which variables?
It isn't clear.

**Comment by data-man** [comment]
> array [0..1023, char]

Here! Again the space! :)

**Comment by kaushalmodi** [comment]
With [this tip][1] from the Nim forums, I learnt that I need to pollute the stack so that the _noinit_ vars don't pick up the "clean stack" values of 0.

Here are my notes for reference: https://scripter.co/notes/nim/#uninitialized-variables

@Araq The requested documentation clarification in [my earlier comment][2] still stands. I spent a lot of time on this trying to understand the `noinit` behavior by _reverse-engineering_. 

[1]: https://forum.nim-lang.org/t/3699#23059
[2]: https://github.com/nim-lang/Nim/issues/7852#issuecomment-390715981

**Comment by zah** [comment]
@kaushalmodi, if you want to know whether `noinit` worked, you just need to look at the generated C code in your `nimcache` folder.

Otherwise, consider that `noinit` is primary about optimization, I'd consider test cases where it didn't kick in as bugs (given that there was no good reason to ignore the explicit instruction).

**Comment by kaushalmodi** [comment]
@zah 

> you just need to look at the generated C code in your nimcache folder.

I haven't coded in C. Nim is the systems programming language that I am deep-diving in i.e. attempting to understand these low level concepts. 

I did try once to try to grok what the generated C looks like.. but a 2-line Nim code generated a 150 line or so C code, and I left my attempt to understand all of that C for a rainy day. 

Now that I think of it, let me generate C code with and without `{.noinit.}`, and let me see what difference I see in there. 

---

But in any case, if `{.noinit.}` is not supposed to work for a case like global vars, shouldn't Nim throw an error if a user tries to *noinit* those? Without any introspection function like `initialized` (just as we have `declared`), or any warning, etc, how would a Nim learner like me realize if that *noinit* actually worked or not?

The manual simply says that this pragma prevents variables from being initialized.. it does **not** clarify that it does so **only for local variables**. 

*Why is it so difficult to get this simple point across to Araq and whoever has the power to fix that in the manual?*

I would have submitted a PR to fix that in the manual in a blink, but knowing that I am still learning this Nim concept, I don't consider myself qualified to do so.

**Comment by GULPF** [comment]
The way I see it, `{.noinit.}` simply tells the compiler that the code doesn't rely on the variable being initialized. Whether the compiler is able to skip the initialization as an optimization depends on what backend is used. For example, JavaScript has no concept of uninitialized variables, so `.{noInit.}` wont result in any optimizations for the JS backend. Even if the variable gets initialized, `{.noInit.}` still "works", since the use case is just to make an additional optimization possible.

I agree that the documentation doesn't make this clear though. The documentation should explicitly say that the compiler is free to initialize variables with `{.noinit.}` anyway, and perhaps mention global variables as an example (when using the C backend).

**Comment by GULPF** [comment]
> I haven't coded in C. Nim is the systems programming language that I am deep-diving in i.e. attempting to understand these low level concepts.
>
> I did try once to try to grok what the generated C looks like.. but a 2-line Nim code generated a 150 line or so C code, and I left my attempt to understand all of that C for a rainy day.

You don't need to understand everything, just search for the variable your interested in. The name will be mangled, but it will be prefixed with the original name. Also, unless you need to see the generated debug code, always use `-d:release` when inspecting the generated code. It makes it much more readable.
